src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
             syscall_error(Errno::ENOENT, "chdir", "the directory referred to in path does not exist")
         }
     }
+
+    //------------------------------------DUP & DUP2 SYSCALLS------------------------------------
+
+    pub fn dup_syscall(&self, fd: i32, start_desc: Option<i32>) -> i32 {
+        let mut fdtable = self.filedescriptortable.write().unwrap();
+
+        let start_fd = match start_desc {
+            Some(start_desc) => start_desc,
+            None => STARTINGFD,
+        };
+
+        //checking whether the fd exists in the file table and is higher than the starting file descriptor
+        if let Some(_) = fdtable.get(&fd) {
+            if fd >= STARTINGFD {
+                return syscall_error(Errno::EBADF, "dup", "provided file descriptor is out of range");
+            }
+
+            //error below may need to be changed -- called if error getting file descriptor
+            let nextfd = if let Some(fd) = self.get_next_fd(Some(start_fd), Some(&fdtable)) {fd} else {return syscall_error(Errno::ENFILE, "dup_syscall", "no available file descriptor number could be found");};
+            return Self::_dup2_helper(&self, fd, nextfd, Some(&mut fdtable))
+        } else {
+            return syscall_error(Errno::EBADF, "dup", "file descriptor not found")
+        }
+    }
+
+    pub fn dup2_syscall(&self, oldfd: i32, newfd: i32) -> i32{
+        let mut fdtable = self.filedescriptortable.write().unwrap();
+
+        if let Some(_) = fdtable.get(&oldfd) {
+            return Self::_dup2_helper(&self, oldfd, newfd, Some(&mut fdtable));
+        } else {
+            return syscall_error(Errno::EBADF, "dup2","Invalid old file descriptor.");
+        }
+    }
+
+    pub fn _dup2_helper(&self, oldfd: i32, newfd: i32, fdtable_lock: Option<&mut FdTable>) -> i32 {
+        let mut writer;
+        let fdtable = if let Some(fdtb) = fdtable_lock {fdtb} else {
+            writer = self.filedescriptortable.write().unwrap(); 
+            &mut writer
+        };
+        
+        //checking if the new fd is out of range
+        if newfd >= MAXFD || newfd <= STARTINGFD {
+            return syscall_error(Errno::EBADF, "dup or dup2", "provided file descriptor is out of range");
+        }
+
+        //if the file descriptors are equal, return the new one
+        if newfd == oldfd {
+            return newfd;
+        }
+
+        //need to add close helper and reference
+        match fdtable.get(&newfd) {
+            Some(_) => {return Self::_close_helper(&self, newfd, Some(&fdtable), false);},
+            None => {
+                fdtable.insert(newfd, fdtable.get(&oldfd).unwrap().clone());
+            } 
+        }
```

Comment:
The logic here seems wrong to me, even if you have to close a file descriptor previously present at the new fd, you should still later insert the duplicated fd.